sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala

```diff
@@ -228,6 +232,95 @@ case class FilterExec(condition: Expression, child: SparkPlan)
   override def outputPartitioning: Partitioning = child.outputPartitioning
 }
 
+/** Physical plan for RecursiveTable. */
+case class RecursiveTableExec(
+    name: String,
+    anchorTerms: Seq[SparkPlan],
+    recursiveTerms: Seq[LogicalPlan],
+    limit: Option[Long]) extends SparkPlan {
+  override def children: Seq[SparkPlan] = anchorTerms
+
+  override def output: Seq[Attribute] = anchorTerms.head.output.map(_.withNullability(true))
+
+  override def simpleString(maxFields: Int): String =
+    s"RecursiveTable $name${limit.map(", " + _).getOrElse("")}"
+
+  override def innerChildren: Seq[LogicalPlan] = recursiveTerms
+
+  override protected def doExecute(): RDD[InternalRow] = {
+    val prevIterationRDDs = ArrayBuffer.empty[RDD[InternalRow]]
+    var prevIterationCount = 0L
+    val anchorTermsIterator = anchorTerms.iterator
+    while (anchorTermsIterator.hasNext && limit.forall(_ > prevIterationCount)) {
+      val anchorTerm = anchorTermsIterator.next()
+
+      val rdd = anchorTerm.execute().map(_.copy()).cache()
```

Comment:
I think we should make the cache level configurable

Reply:
good idea, I've added a new config `spark.sql.cte.recursion.cache.storageLevel`